File: src/main/java/com/qtech/bubbles/screen/GameOverScreen.java
package com.qtech.bubbles.screen;

import com.qtech.bubbles.BubbleBlaster;
import com.qtech.bubbles.core.utils.categories.GraphicsUtils;
import com.qtech.bubbles.data.GlobalSaveData;
import com.qtech.bubbles.event.bus.EventBus;

import java.awt.*;

public class GameOverScreen extends Screen {
    private final boolean isHighScore;
    private final Font titleFont = new Font(BubbleBlaster.getInstance().getSansFontName(), Font.BOLD, 64);
    private final Font descriptionFont = new Font(BubbleBlaster.getInstance().getSansFontName(), Font.BOLD, 14);
    private final Font scoreFont = new Font(BubbleBlaster.getInstance().getSansFontName(), Font.BOLD, 32);
    private long score;
    private EventBus.Handler binding;

    public GameOverScreen(long score) {
        this.score = score;

        GlobalSaveData globalSaveData = GlobalSaveData.instance();
        if (!globalSaveData.isLoaded()) {
            if (!globalSaveData.isCreated()) {
                globalSaveData.create();
            }
            globalSaveData.load();
        }

        isHighScore = globalSaveData.getHighScore() < score;

        if (isHighScore) {
            globalSaveData.setHighScore(score, System.currentTimeMillis());
            globalSaveData.dump();
        }

        this.score = score;
    }

    @Override
    public void init() {
        bindEvents();
    }

    @Override
    public boolean onClose(Screen to) {
        unbindEvents();

        return super.onClose(to);
    }

    @Override
    public void render(BubbleBlaster game, Graphics2D gg) {
        LinearGradientPaint gradient = new LinearGradientPaint(0, 0, 0, game.getScaledHeight(), new float[]{0.0f, 0.4f}, new Color[]{new Color(0, 0, 224), new Color(0, 0, 128)});
        Paint old = gg.getPaint();
        gg.setPaint(gradient);
        gg.fillRect(0, 0, game.getScaledWidth(), game.getScaledHeight());
        gg.setPaint(old);
    }

    @Override
    public void renderGUI(BubbleBlaster game, Graphics2D gg) {
        gg.setColor(Color.white);

        if (isHighScore) {
            GraphicsUtils.drawCenteredString(gg, "Congratulations!", new Rectangle(0, 120, game.getScaledWidth(), 64), titleFont);
            GraphicsUtils.drawCenteredString(gg, "You beat your high-score!", new Rectangle(0, 184, game.getScaledWidth(), 64), descriptionFont);
        }

        GraphicsUtils.drawCenteredString(gg, Long.toString(score), new Rectangle(0, 248, game.getScaledWidth(), 64), scoreFont);
    }

    public boolean isHighScore() {
        return isHighScore;
    }
}

File: src/main/kotlin/com/qtech/bubbles/screen/GameOverScreen.kt
package com.qtech.bubbles.screen

import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.common.screen.Screen
import com.qtech.bubbles.core.utils.categories.GraphicsUtils.drawCenteredString
import com.qtech.bubbles.data.GlobalSaveData
import com.qtech.bubbles.graphics.GraphicsProcessor
import java.awt.Color
import java.awt.Font
import java.awt.LinearGradientPaint
import java.awt.Rectangle

class GameOverScreen(private var score: Long) : Screen() {
    val isHighScore: Boolean
    private val titleFont = Font(BubbleBlaster.instance.sansFontName, Font.BOLD, 64)
    private val descriptionFont = Font(BubbleBlaster.instance.sansFontName, Font.BOLD, 14)
    private val scoreFont = Font(BubbleBlaster.instance.sansFontName, Font.BOLD, 32)
    override fun init() {
        bindEvents()
    }

    override fun onClose(to: Screen?): Boolean {
        unbindEvents()
        return super.onClose(to)
    }

    override fun render(game: BubbleBlaster, gp: GraphicsProcessor) {
        val gradient = LinearGradientPaint(0F, 0F, 0F, game.height.toFloat(), floatArrayOf(0.0f, 0.4f), arrayOf(Color(0, 0, 224), Color(0, 0, 128)))
        val old = gp.paint()
        gp.paint(gradient)
            .rectF(0, 0, game.width, game.height)
            .paint(old)
    }

    override fun renderGUI(game: BubbleBlaster, gg: GraphicsProcessor) {
        gg.color(Color.white)
        if (isHighScore) {
            drawCenteredString(gg, "Congratulations!", Rectangle(0, 120, game.width, 64), titleFont)
            drawCenteredString(gg, "You beat your high-score!", Rectangle(0, 184, game.width, 64), descriptionFont)
        }
        drawCenteredString(gg, score.toString(), Rectangle(0, 248, game.width, 64), scoreFont)
    }

    init {
        val globalSaveData: GlobalSaveData = GlobalSaveData.instance()
        if (!globalSaveData.isLoaded) {
            if (!globalSaveData.isCreated) {
                globalSaveData.create()
            }
            globalSaveData.load()
        }
        isHighScore = globalSaveData.highScore < score
        if (isHighScore) {
            globalSaveData.setHighScore(score.toDouble(), System.currentTimeMillis())
            globalSaveData.dump()
        }
    }
}